Python/ceval_gil.c

```diff
@@ -57,113 +57,60 @@
 #define _Py_atomic_load_relaxed_int32(ATOMIC_VAL) _Py_atomic_load_relaxed(ATOMIC_VAL)
 #endif
 
-/* This can set eval_breaker to 0 even though gil_drop_request became
-   1.  We believe this is all right because the eval loop will release
-   the GIL eventually anyway. */
+
 static inline void
-COMPUTE_EVAL_BREAKER(PyInterpreterState *interp,
-                     struct _ceval_runtime_state *ceval,
-                     struct _ceval_state *ceval2)
+update_eval_breaker_from_thread(PyInterpreterState *interp, PyThreadState *tstate)
 {
-    _Py_atomic_store_relaxed(&ceval2->eval_breaker,
-        _Py_atomic_load_relaxed_int32(&ceval2->gil_drop_request)
-        | (_Py_atomic_load_relaxed_int32(&ceval->signals_pending)
-           && _Py_ThreadCanHandleSignals(interp))
-        | (_Py_atomic_load_relaxed_int32(&ceval2->pending.calls_to_do))
-        | (_Py_IsMainThread() && _Py_IsMainInterpreter(interp)
-           &&_Py_atomic_load_relaxed_int32(&ceval->pending_mainthread.calls_to_do))
-        | ceval2->pending.async_exc
-        | _Py_atomic_load_relaxed_int32(&ceval2->gc_scheduled));
+    if (tstate == NULL) {
+        return;
+    }
+    int32_t calls_to_do;
+    if (_Py_ThreadCanHandleSignals(interp)) {
+        calls_to_do = _Py_atomic_load_int32_relaxed(
+            &_PyRuntime.ceval.pending_mainthread.calls_to_do);
```

Comment:
Isn't this fully covered by the places we call `SIGNAL_PENDING_CALLS()`?

Reply:
Not in this case.

When running a non-main thread with `pending_mainthread.calls_to_do` we don't want  `_PY_CALLS_TO_DO_BIT` set, to avoid constantly calling `_Py_HandlePending`.
So when we switch to the main thread, we need to set  `_PY_CALLS_TO_DO_BIT` if `pending_mainthread.calls_to_do`

Reply:
Similarly to https://github.com/python/cpython/pull/109846#discussion_r1341646682, why does acquiring the GIL require that we reproduce what was already done via `Py_AddPendingCall()`?

Is it because we always flip the bit in `_make_pending_calls()`, even if we're not in the main thread but there are global pending calls waiting.  That seems like we could use a separate bit for the global pending calls.

Is i because we don't want to interrupt all the other threads while waiting for the main thread to handle its pending calls?  I suppose the current approach sort of addresses that, but (as indicated in https://github.com/python/cpython/pull/109846#discussion_r1341646682) that means we put off those extraneous interruptions only until the main thread takes the GIL again.  Then the interruptions resume until the main thread handles its pending calls.  That feels very jumpy.  ISTM, a per-thread eval breaker would be a clearer solution.

This all applies to the signal-related code here in `update_eval_breaker_from_thread()` too.

Reply:
I'm not suggesting that we change things up in this PR.  Rather, **if** a less jumpy approach makes sense then we could do it in a follow-up PR.